File: src/contests/cg/puzzle/haskell/telephone-numbers.hs
-- 🎮 CodinGame Puzzle - telephone-numbers
-- https://www.codingame.com/training/medium/telephone-numbers

import Data.List (inits)
import qualified Data.Set as Set

-- Each distinct non-empty prefix is exactly one trie node (= one cable unit).
main :: IO ()
main = do
  n <- readLn :: IO Int
  ls <- mapM (const getLine) [1 .. n]
  let prefixes = Set.fromList (concatMap (filter (not . null) . inits) ls)
  print (Set.size prefixes)

File: src/contests/cg/puzzle/clojure/telephone-numbers.clj
;; 🎮 CodinGame Puzzle - telephone-numbers
;; https://www.codingame.com/training/medium/telephone-numbers

(ns Solution (:gen-class))

;; Each distinct non-empty prefix is exactly one trie node (= one cable unit).
(defn -main [& _]
  (let [n (Integer/parseInt (.trim (read-line)))
        prefixes (reduce
                  (fn [s _]
                    (let [num (.trim (read-line))]
                      (reduce (fn [s2 i] (conj s2 (subs num 0 (inc i))))
                              s (range (count num)))))
                  #{} (range n))]
    (println (count prefixes)))
  (flush))
